src/lib.rs

```diff
@@ -1528,13 +1553,13 @@ pub mod __rt {
 
     impl<'b, T: ?Sized> Drop for Ref<'b, T> {
         fn drop(&mut self) {
-            self.borrow.set(self.borrow.get() - 1);
+            self.borrow.fetch_sub(1, Ordering::SeqCst);
```

Comment:
Is there a reason we need `SeqCst` instead of `Relaxed` everywhere?

Reply:
Digging a bit deeper, it looks like [today, all atomic accesses in WebAssembly are SeqCst](https://github.com/WebAssembly/threads/blob/master/proposals/threads/Overview.md#atomic-memory-accesses):

> Currently all atomic memory access instructions are [sequentially consistent](https://en.wikipedia.org/wiki/Sequential_consistency). Instructions with other memory orderings may be provided in the future.

But what's the "right" ordering that's maximally correct and efficient, and remains so as we add weaker orderings to the WebAssembly spec?

I'm pretty sure `Relaxed` is wrong. Compilers are allowed to reorder other memory accesses across a `Relaxed` store or load, so the compiler could generate code that accesses `self.value` before we've confirmed that it's safe to access.

[The Rustonomicon's guide to Atomics](https://doc.rust-lang.org/nomicon/atomics.html) says that you should use `Acquire` to mark the beginning of a critical section and `Release` to mark the end:

> Intuitively, an acquire access ensures that every access after it stays after it. However operations that occur *before* an acquire are free to be reordered to occur *after* it. Similarly, a release access ensures that every access before it stays before it. However operations that occur *after* a release are free to be reordered to occur *before* it.

> Basic use of release-acquire is therefore simple: you acquire a location of memory to begin the critical section, and then release that location to end it.

We can also look at the memory orderings used for Rust's [`std::sync::RwLock`](https://doc.rust-lang.org/src/std/sys/unix/locks/futex_rwlock.rs.html#77-156) implementation. We are essentially rewriting `RwLock::try_read()` and `RwLock::try_write()`, but calling `borrow_fail()` when those functions would return `false`.

```
pub fn try_read(&self) -> bool {
    self.state
        .fetch_update(Acquire, Relaxed, |s| is_read_lockable(s).then(|| s + READ_LOCKED))
        .is_ok()
}

pub fn try_write(&self) -> bool {
    self.state
        .fetch_update(Acquire, Relaxed, |s| is_unlocked(s).then(|| s + WRITE_LOCKED))
        .is_ok()
}

pub unsafe fn read_unlock(&self) {
    let state = self.state.fetch_sub(READ_LOCKED, Release) - READ_LOCKED;

    // It's impossible for a reader to be waiting on a read-locked RwLock,
    // except if there is also a writer waiting.
    debug_assert!(!has_readers_waiting(state) || has_writers_waiting(state));

    // Wake up a writer if we were the last reader and there's a writer waiting.
    if is_unlocked(state) && has_writers_waiting(state) {
        self.wake_writer_or_readers(state);
    }
}

pub unsafe fn write_unlock(&self) {
    let state = self.state.fetch_sub(WRITE_LOCKED, Release) - WRITE_LOCKED;

    debug_assert!(is_unlocked(state));

    if has_writers_waiting(state) || has_readers_waiting(state) {
        self.wake_writer_or_readers(state);
    }
}
```

So I'm going to go ahead and rewrite this to use `Acquire` and `Release` orderings when we want to acquire and release our semaphore, using `RwLock` as a guide. But unfortunately this is all kind of theoretical and the generated wasm is always the same.

Reply:
> I'm pretty sure `Relaxed` is wrong. Compilers are allowed to reorder other memory accesses across a `Relaxed` store or load, so the compiler could generate code that accesses `self.value` before we've confirmed that it's safe to access.

I don't believe this is correct, we don't have a "critical section" per se, we have a check for the returned value, this can't be reordered, because the `if` **has** to happen before it's body is executed.

I don't think `RwLock` is comparable because they have to sync access between multiple threads over a critical section, e.g. [this part](https://doc.rust-lang.org/1.72.0/src/std/sys/unix/locks/futex_rwlock.rs.html#202-218). We don't have that, for us it doesn't really matter who comes first as we don't coordinate state access atomically, we just give value access in certain conditions.